_posts/2018-07-10-react-native-wdio.md

```diff
@@ -0,0 +1,301 @@
+
+```diff
+- services: ['appium'],
++ services: ['sauce'],
++ host: 'ondemand.saucelabs.com',
+- port: 4723,
++ port: 80,
++ user: SAUCE_USERNAME,
++ key: SAUCE_ACCESS_KEY
+```
+
+Use [Platform configurator by Sauce Labs](https://wiki.saucelabs.com/display/DOCS/Platform+Configurator) to update your capabilities if needed.
+
+**Step 3:** Upload your `.app` or `.apk` file to SauceStorage or any other accessible endpoint. Follow the documentation here: [Uploading Mobile Applications to Sauce Storage for Testing](https://wiki.saucelabs.com/display/DOCS/Uploading+Mobile+Applications+to+Sauce+Storage+for+Testing)
+
+**Update App path in capabilities:**
+```diff
+- app: <app path>
++ app: sauce-storage:myapp.zip or app link
+```
+
+**Run the test:**
+```console
+$ wdio wdio.conf.js
+```
+
+Check the [Sauce Labs Dashboard](https://saucelabs.com/beta/dashboard/tests) to make sure the test has run. The WebDriverIO Sauce service automatically sets test labels and results.
+
+
+## Part 3: Running on Real Devices
+Sauce Labs also provides real device a testing solution through the TestObject platform. With some minor changes to the above `wdio.conf`, we can run UI tests on real devices.
+
+First, create a project in the [TestObject dashboard](https://app.testobject.com/) and upload your `.ipa` or `.apk` file.
+
+There is no official WebDriverIO TestObject service and at the time of writing this blog, Sauce Labs and TestObject do not share the same API calls.
+
+**Update `wdio.conf` file:**
+```diff
+- services: ['sauce'],
++ protocol: 'https',
++ host: 'us1.appium.testobject.com',
+- port: '80',
++ port: '443',
++ path: '/api/appium/wd/hub',
+capabilities: {
+  ...
+-  app: 'sauce-storage:myapp.zip'
++  testobject_api_key: <TESTOBJECT_ACCESS_KEY>,
++  testobject_app_id: <APP_ID>,
+}
+```
+
+Refer to [Appium Capabilities for Real Device Testing](https://wiki.saucelabs.com/display/DOCS/Appium+Capabilities+for+Real+Device+Testing) to update your capabilities if needed.
+
+**Run the UI test:**
+
+```console
+$ wdio wdio.conf
+```
+
+Check the [TestObject Dashboard](https://app.testobject.com/) to make sure the test has run. WebDriverIO does not update test labels and results for TestObject. Here are some references that can help update the test results: <https://github.com/pizzasaurusrex/TestObject> or <https://gist.github.com/rajapanidepu/0e8c0f89671a8a563a7463f8c1ff0413>
+
+
+## Part 4: Timeouts
+When we started to write UI tests and run them as part of CICD, timeouts played a very important role in making sure UI tests were stable. We observed that our tests were failing for reasons outside of the test code, such as the device not being ready or the app installation taking a long time.
+
+There are various timeouts in play here at each level of tech stack: WebDriverIO, Mocha/Jasmine, Appium, and Sauce Labs/TestObject. In our case, simple WebDriverIO tests kept failing intermittently until we completely understood the timeouts and tweaked them.
```

Comment:
Can you offer guidance on what specifically needs tweaking (why wouldn't the default by the tweaked value then?) and why?

Reply:
Sure. More info can be added in here.